### ISSUE TITLE
Feature Request: RAII Lock Guard for Lock/Mutex Library

### ISSUE THREAD

**Issue by LouisJenkinsCS** [chapel-lang/chapel#10341]
I would like to request something similar to C++'s [std::lock_guard](https://en.cppreference.com/w/cpp/thread/lock_guard) where the user can define the scope in which they want to acquire a specific lock. The lock guards should work seamlessly with the locks offered by the Lock/Mutex library. Example usage that I would appreciate would be...

```chpl
var lock : SpinLock;
// Automatically acquires lock
sync lock {
   // ...
}
// Automatically releases lock
```

Something like this would be great. I know the `sync` statement, according to the language specification, is used to wait on spawned tasks to await their termination, but I wonder if it can be extended a bit to work like shown above. If that can't work,  perhaps it could be a specific `record` that releases the lock once it goes out of scope? I remember doing this before in issue #6912 but there were other issues with memory management that prevented me from doing so. @ronawho since you suggested building the Lock/Mutex library in issue #9881

**Comment by mppf** [comment]
We already have RAII variables if they are records.

I.e. this could already work (with module code implementing the lock and lock guard types):

``` chapel
...
{
   var lock = new HeldLock(GlobalLock);
} // lock is released here
...
```

**Comment by bradcray** [comment]
Hi @LouisJenkinsCS -- I'm wondering whether @mppf's response above is sufficient to close this issue in favor of prioritizing other ones?
